src/praisonai-agents/praisonaiagents/tools/file_tools.py

```
@@ -116,16 +133,41 @@ def write_file(self, filepath: str, content: str, encoding: str = 'utf-8') -> bo
+            # Serialise concurrent writes/edits on this canonical path.
+            with _file_locks.get_lock(safe_path):
+                # Automatic staleness guard for existing files.
+                if not force and os.path.exists(safe_path):
+                    recorded = _file_locks.get_read_hash(safe_path)
+                    if recorded is not None:
+                        try:
+                            with open(safe_path, 'r', encoding=encoding) as f:
+                                on_disk = f.read()
+                            current = hashlib.sha256(
+                                on_disk.encode('utf-8')).hexdigest()
+                        except Exception:
```

Ruff (0.15.18)

[warning] Do not catch blind exception: `Exception`

(BLE001)